### PR TITLE
KEYCLOAK-14974 Map group storage provider

### PR DESCRIPTION
**Pull request by mhajas**

### CODE REVIEW

**Comment by mhajas** [comment]
@hmlnarik @vramik This PR should be ready for review. There is just some `SocketTimeout` failure in `Keycloak.X` tests which is probably not caused by these changes.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:561]: I believe this method will soon become overused from other parts as well (e.g. users). Could you clearly mark by comments that this particular command is there due to GroupProvider? This will become useful once we work on https://issues.redhat.com/browse/KEYCLOAK-14750.

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupAdapter.java
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/group/MapGroupAdapter.java:147]: Extract `app.getId()` to a local variable before the stream.

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:276]: Extract `role.getId()`

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:277]: Extract `groupEntity.getId().toString()`

File: model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java:806]:
```suggestion
        groupsIds.add(group.getId());
```

File: model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java:815]:
```suggestion
        if (groupIds.remove(group.getId())) {
```

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Good job, @mhajas, thanks! Please find the review comment inline

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java
Comment (mhajas) [reply model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java:966]: Thank you for pointing this out. It just looked suspicious to me, as `getRealmRolesStream(RealmModel realm)` is using this, I think also this method should check it.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:484]: This something we should probably unify across the whole codebase because I saw more approaches to this [(for example).](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java#L460). Maybe we should add this to the cleanup task.

I did it this way because it is used here, but I will fix: https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/services/resources/admin/GroupsResource.java#L84

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:490]: Sure

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:277]: Do you mean to add another `map` function call, with changing `groupEntity` stream to String stream?

**Review by vramik — COMMENTED** [review]
@mhajas thanks for the PR, It looks very good to me. Few comments inline.

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:299]: Ah, sorry, ignore this one. :-)

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (hmlnarik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java]: Absolutely agree, we should follow single approach to these parameters. Thanks for fixing.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (hmlnarik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:585]: I meant really only to mark the lines within the method like 
```suggestion
        // GroupProvider interface method implementation starts here
        em.createNamedQuery("deleteGroupRoleMappingsByRole").setParameter("roleId", role.getId()).executeUpdate();
        // GroupProvider interface method implementation ends here
```
Apologies for not being clear before

File: model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java
Comment (vramik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java:797]: Is this check here useful/necessary? `realm.getDefaultGroupIds()` never returns null and if I understand it correctly calling `.map` on empty stream is not expensive operation. @mhajas, @hmlnarik wdyt?

File: model/jpa/src/main/resources/META-INF/jpa-changelog-12.0.0.xml
Comment (vramik) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-12.0.0.xml:20]: There will be probably another required changes in this changeSet which will be related to role work, could we rename `id` to something less specific? Maybe `map-remove-ri`?

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:98]: nit:
```suggestion
                .map(tx::getUpdated)    // If the group has been removed, tx.get will return null, otherwise it will return me.getValue()
```

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:216]: I don't understand the comment, could you please explain?

File: server-spi/src/main/java/org/keycloak/models/GroupModel.java
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/models/GroupModel.java:40]: Sound good, added

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:585]: No apologies needed, I should have got this. Fixed

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:216]: The thing here is, that in the database we have a constraint that checks there are no two groups with the same name and parent. Here, we need to do it manually.

File: model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java
Comment (mhajas) [inline model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java:411]: @hmlnarik @vramik Should we have a count query here? This would mean we need another count method in GroupProvider which is matching this method: `session.groups().getGroupsStream(realm, closing(createGetGroupsQuery().getResultStream()), search, first, max)`

**Comment by mhajas** [comment]
@hmlnarik @vramik I think I addressed all your suggestions in `Addressing comments` commit. Could you have another look, please?

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:216]: This note has to be present in `GroupProvider.createGroup` JavaDocs

File: model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java
Comment (hmlnarik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java]: The `RealmModel.getDefaultGroupIds()` has to state that it `never returns null` to have this as a contract.

File: model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java
Comment (hmlnarik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java:411]: Yes

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmEntity.java
Comment (vramik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmEntity.java:167]:
```suggestion
    protected Set<String> defaultGroupIds;
```

Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmEntity.java]: Done

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:439]: There need to be two queries, one with and one without `search` parameter, for performance reasons to prevent seq scan in case of no search

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:458]: same as above

File: services/src/main/java/org/keycloak/services/resources/admin/GroupResource.java
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/services/resources/admin/GroupResource.java:149]: This change needs to be accompanied with a change that interprets 409 correctly in admin UI

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:458]: I believe there is no need for such a function. We should probably throw an exception here if the search is `null`. It doesn't make sense to search for a count of groups with ids (we can just use the size of ids list). In the adapter layer, we should make sure no such call is made.

Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:458]: Looked at it again and an exception is probably not the best, but we can just call `ids.map(id -> session.groups().getGroupById(realm, id))`, it should be probably faster, WDYT?

File: services/src/main/java/org/keycloak/services/resources/admin/GroupResource.java
Comment (mhajas) [reply services/src/main/java/org/keycloak/services/resources/admin/GroupResource.java:149]: I believe it should work the same. Look at the next line, there was:
```return ErrorResponse.exists("Parent already contains subgroup named '" + rep.getName() + "'");```

which should be the same 409, but I will check.

**Review by vramik — CHANGES_REQUESTED** [review]
Thank you @mhajas, lgtm, If I see it correctly there are two remaining items from previous discussion. Other than that as far as I'm concerned it's ready. Great job.

- https://github.com/keycloak/keycloak/pull/7387#discussion_r486957618
- https://github.com/keycloak/keycloak/pull/7387#discussion_r486956508

File: model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java]: The `getDefaultGroupIds()` here is called on the `RealmEntity` (not public interface) not on the RealmModel. Do we still need to add the `never returns null` claim?

**Comment by mhajas** [comment]
> Thank you @mhajas, lgtm, If I see it correctly there are two remaining items from previous discussion. Other than that as far as I'm concerned it's ready. Great job.
> 
> * [#7387 (comment)](https://github.com/keycloak/keycloak/pull/7387#discussion_r486957618)
I commented it there.

> * [#7387 (comment)](https://github.com/keycloak/keycloak/pull/7387#discussion_r486956508)
This should be already addressed. I added `@throws` to Javadoc of all createGroup variants.

File: server-spi/src/main/java/org/keycloak/models/UserModel.java
Comment (hmlnarik) [inline server-spi/src/main/java/org/keycloak/models/UserModel.java:143]: Is it case sensitive or insensitive? Can it be null?

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/CachedUser.java
Comment (hmlnarik) [inline model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/CachedUser.java:31]: Is this necessary?

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java]: I see, ok I fixed

**Comment by mhajas** [comment]
@hmlnarik @vramik Pipeline runs:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2143/
https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/1729/

**Review by vramik — DISMISSED** [review]

**Comment by hmlnarik** [comment]
LGTM, thanks @mhajas ! One last thing before getting merged - could you please squash the commits now?

**Comment by mhajas** [comment]
@hmlnarik done

**Review by hmlnarik — APPROVED** [review]

**Review by vramik — APPROVED** [review]

**Pull request merged**
